spec.bs

```diff
@@ -1226,6 +1221,273 @@ On the other hand, methods for getting data from the [=shared storage database=]
+
+    add the step
+
+    1. [=Append or modify a Shared-Storage-Writable request header=] for |httpRequest|.
+  </div>
+
+### Modification to [=HTTP fetch=] Algorithm ### {#mod-http-fetch}
+
+  <div algorithm="modification-to-http-fetch">
+    Modify the [=HTTP fetch=] algorithm as follows:
+
+    Before the step
+
+    > If |internalResponse|'s [=response/status=] is a [=redirect status=]: ...
+
+    add the step
+
+    1. [=Handle a Shared-Storage-Write response=], given [=/response=] |internalResponse| and [=/request=] <var ignore=''>request</var> as input.
+  </div>
+
+## Shared Storage HTTP Headers ## {#headers}
+
+### [:Shared-Storage-Writable:] Request Header ### {#request-header}
+
+  This specification defines a <dfn http-header>Shared-Storage-Writable</dfn> HTTP [=/request=] [=header=].
+
+  The [:Shared-Storage-Writable:] [=/request=] [=header=] is a [=Structured Header=] whose value must be a [=structured header/Boolean=].
+
+  When the result of running [=get a structured field value=] for "[:Shared-Storage-Writable:]", "`item`", and [=/request=]'s [=request/header list=] is true, then the [=/response=] to this [=/request=] will be eligible to write to [=shared storage=].
+
+### [:Shared-Storage-Write:] Response Header ### {#response-header}
+
+  This specification defines a <dfn http-header>Shared-Storage-Write</dfn> HTTP [=/response=] [=header=].
+
+  The [:Shared-Storage-Write:] [=/response=] [=header=] is a [=Structured Header=] whose value must be a [=structured header/List=], for which the [=structured header/parameters|parameterized=] [=structured header/items=] can have one of the following values, expressed as [=structured header/Token=], [=structured header/String=], or [=structured header/Byte Sequence=]:
+
+    * `set`
+        - required [=structured header/parameters=]:
+            * key ([=structured header/Token=], [=structured header/String=], or [=structured header/Byte Sequence=])
+            * value ([=structured header/Token=], [=structured header/String=], or [=structured header/Byte Sequence=])
+        - optional [=structured header/parameters|parameter=]:
+            * ignore_if_present ([=structured header/Boolean=])
+    * `append`
+        - required [=structured header/parameters=]:
+            * key ([=structured header/Token=], [=structured header/String=], or [=structured header/Byte Sequence=])
+            * value ([=structured header/Token=], [=structured header/String=], or [=structured header/Byte Sequence=])
+    * `delete`
+        - required [=structured header/parameters|parameter=]:
+            * key ([=structured header/Token=], [=structured header/String=], or [=structured header/Byte Sequence=])
+    * `clear`
+
+    Note: We allow [=structured header/Byte Sequences=] in order to accommodate Unicode keys and values. Any [=structured header/Byte Sequence=] will be assumed to be [=UTF-8 encoded=] and will fail to parse otherwise.
```

Comment:
Consider using https://httpwg.org/http-extensions/draft-ietf-httpbis-sfbis.html#name-display-strings (new in the latest draft of Structured Fields, and not in an RFC yet) instead of byte sequences to store Unicode. Note the disagreements about adding them in https://github.com/httpwg/http-extensions/pull/2494 and https://lists.w3.org/Archives/Public/ietf-http-wg/2023AprJun/thread.html#msg177. It might be worth mailing the HTTPWG with this use case to get their opinions on how to encode what you want to encode.

Reply:
Thanks for bring to my attention. 

Oh wow, display strings seem very unsafe, and I agree with those asking why they are needed given that we already have the safer byte sequences...

I can discuss with my team, but I would definitely want to have an in-depth conversation with some Chromium network security people before switching over to display strings, if we do that. I can put this on our todo list for potential changes that need research first, in case display strings are adopted...

Reply:
Added issue #117 .